doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -99,14 +99,26 @@ Keywords
   The following character sequences are reserved keywords that cannot be
   used as identifiers::
 
-    _ Axiom CoFixpoint Definition Fixpoint Hypothesis IF Parameter Prop
-    SProp Set Theorem Type Variable as at by cofix discriminated else
-    end exists exists2 fix for forall fun if in lazymatch let match
-    multimatch return then using where with
+    _ Axiom CoFixpoint Definition Fixpoint Hypothesis Parameter Prop
+    SProp Set Theorem Type Variable as at cofix else
+    end fix for forall fun if in let match
+    return then where with
+
+  The following identifiers are not reserved keywords but are
```

Comment:
```suggestion
  The following identifiers are reserved and may not be used as the name of an Ltac function or as part of the definition of a tactic notation.
```

Reply:
Let's remember your formulation which I suggest we eventually put in the Ltac and Tactic Notations chapters. I prefer to keep to the Gallina chapter independent of what other features provide.

Reply:
I agree with keeping the Gallina chapter independent, but you could still add something to say that their use is restricted at the beginning of sentences, WDYT?

Reply:
> I prefer to keep to the Gallina chapter independent of what other features provide.

What's the benefit of keeping things independent?  I think one of the major problems with the doc is that information is scattered rather than integrated.  Mentioning the reserved identifiers here at all is inconsistent with keeping the chapters independent.  (Look at our chapter titles--we seem to have a pattern of creating new chapters instead of integrating material in a sensible way.)

Except for optionally-loaded plugins, Coq should be considered a unified product.  Nor sure there be fiefdoms within it and its code (if this is, perhaps, one of your considerations).

Reply:
I agree that "one of the major problems with the doc is that information is scattered rather than integrated" and this is why I think in particular that the move of the grammar of sorts, binders, records, projections in the Gallina chapter is good.

But I think that a few conceptual distinctions are important. In particular, Gallina, the tactics, and the libraries are well differentiated concepts. Gallina refers to the logic, to the specification language. It is pretty stable and rather well-delimited even if it is regularly extended with new features (universe polymorphism, `SProp`, implicit arguments, negative types, ...). 

Tactics is not only of a different nature. It is also a moving/extensible target. There are Ltac1, Ltac2, Mtac2, vanilla basic bricks, ssreflect basic bricks, and many cool ideas of new tactics defined here and there, e.g. in TLC. We can much better than what we have now and I don't think it is worth to place the existing tactics at the heart of the documentation.

Regarding libraries (and the keywords which come with), we could somehow consider the initial state, and in particular the logical concepts `Logic.v` (`/\`, `\/`, `exists`, `=`, ...), basic datatypes `Datatypes` (`list`, `option`, `sum`, `prod`, `nat`, ...) and basic sets `Specifs` (`{ x | P }`, `{ x & P }`, ...) almost as a stable part of the system. So, this part could be documented much earlier than it is now. There are still a couple of problems to solve though before stabilizing it and placing it at the heart of the documentation: there is no good notation for the injections of a sum, no good notation for the various projections (for instance, I found `proj1_sig` impossible to remember and inconsistent with the name of other projections). So, I would support documenting the initial state much earlier than it is now, but not w/o first proceeding with a few changes to make it unobjectable.

Reply:
To further support Hugo's point, the extensibility of Coq is really one of its strength, as it has allowed people from various horizons to propose alternative solutions to many hard problems. So no, you should not consider Coq a unified product, as that would be counter-productive (and by this, I do not mean that we should not attempt to standardize things, standardization is an important step after creativity has been expressed through many alternative solutions and understanding has progressed).

This is independent of the fact that the current way of adding stuff to the manual has often been by adding new chapters in the Addendum part. This too is wrong: extended pattern-matching, implicit coercions, canonical structures, typeclasses, generalized rewriting, parallel proof processing, polymorphism, SProp, and even Program, should not be documented as independent chapters because they are not independent extensions.

Reply:
Let me say what I'd do more simply:  If there's no way for a user to avoid having some keyword defined (e.g. it's used in the syntax of some always-loaded tactic) I would include it in a the general section on grammars (which happens to be in the Gallina chapter for now, but probably belongs elsewhere).  That makes it easy to find those without looking at multiple pages in the manual.  If and when we change the code in some future release so that's no longer the case, then at that time we should update the documentation to reflect that.  I think the fact that Gallina is more mature/stable than `Ltac1`, ... doesn't really matter--it's whether it's user configurable.  It think it's fine to repeat that info in other places, such as the tactics chapter.

I would also include info on keywords defined by commonly-loaded plugins in a central section.  Seems to me Ltac is one of those.  SSR?  Not sure.  (Off topic: perhaps we should encourage plugins to limit the number of new keywords they define.)

> Regarding libraries (and the keywords which come with), we could somehow consider the initial state, and in particular the logical concepts `Logic.v` (`/\`, `\/`, `exists`, `=`, ...), basic datatypes `Datatypes` (`list`, `option`, `sum`, `prod`, `nat`, ...) and basic sets `Specifs` (`{ x | P }`, `{ x & P }`, ...) almost as a stable part of the system. So, this part could be documented much earlier than it is now. 

I agree, we should document these better.

> a more scalable approach would simply to provide a command `Print Keywords` which tells what are the keywords defined in the current environment.a more scalable approach would simply to provide a command Print Keywords which tells what are the keywords defined in the current environment.

Yes, good idea.  And make sure we have good commands for showing defined notations, too.